File: 349-intersection-of-two-arrays/intersection-of-two-arrays.go
func intersection(nums1 []int, nums2 []int) []int {
	nums1Map := make(map[int]struct{})
	for _, num := range nums1 {
		nums1Map[num] = struct{}{}
	}

	resultMap := make(map[int]struct{})
	for _, num := range nums2 {
		if _, exists := nums1Map[num]; exists {
			resultMap[num] = struct{}{}
		}
	}

	resultSlice := make([]int, 0, len(resultMap))
	for num := range resultMap {
		resultSlice = append(resultSlice, num)
	}

	return resultSlice
}

File: 349-intersection-of-two-arrays/intersection-of-two-arrays.rs
use std::collections::HashSet;

impl Solution {
    pub fn intersection(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let nums2_set = nums2.into_iter().collect::<HashSet<i32>>();
        nums1
            .into_iter()
            .filter(|e| nums2_set.contains(e))
            .collect::<HashSet<i32>>()
            .into_iter()
            .collect()
    }
}